File: problems/2095.delete-the-middle-node-of-a-linked-list.cpp
/*
 * @lc app=leetcode id=2095 lang=cpp
 *
 * [2095] Delete the Middle Node of a Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  ListNode *deleteMiddle(ListNode *head) {
    int n = 0;
    ListNode *node = head;

    while (node != nullptr) {
      node = node->next;
      n++;
    }

    if (n == 1) {
      return nullptr;
    }

    int mid = n / 2;
    ListNode* cur = head;
    for (int i = 0; i < mid-1; i++) {
      cur = cur->next;
    }
    cur->next = cur->next->next;

    return head;
  }
};
// @lc code=end

File: python/2095.delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from math import ceil, floor


class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return None

        count = 1
        node = head.next
        while node is not None:
            node = node.next
            count += 1

        mid = count // 2

        prev = ListNode(0, head)
        for _ in range(mid):
            prev = prev.next

        prev.next = prev.next.next

        return head
